include/crypto/event.h

```diff
@@ -0,0 +1,378 @@
+     */
+
+    /* A numeric event identity, defined by the using functionality */
+    uint32_t type;
+    /* A context relevant for the using functionality */
+    void *ctx;
+
+    /*
+     * Optional data (may be NULL).
+     *
+     * These fields are considered mutable, in so far that an event handler
+     * may "steal" them, and replace them with NULL.
+     */
+
+    /* Identifying material */
+    const void *identifiers;
+    /* Data carried by the event */
+    size_t payload_size;
+    void *payload;
+
+    /*
+     * A destructor for |identifiers| and |payload|, if necessary.
+     *
+     * This field is considered immutable until the event is destroyed.
+     */
+    ossl_event_destructor_fn *destructor;
+
+    /*
+     * Flags, only relevant for the event functionality itself.
+     */
+    unsigned int flag_dynamic : 1;
+};
+
+/*
+ * The event queue is expected to be very long lived, even thought it may be
+ * dynamically allocated.  Subsribers are expected to return values that
+ * express if the event it received was used or not.
+ */
+enum ossl_event_subscriber_returns {
+    /* The subscriber encountered an error */
+    ossl_event_error = -1,
+    /* The subscriber didn't use this event */
+    ossl_event_pass = 0,
+    /* The subscriber used this event */
+    ossl_event_handled = 1,
+};
+
+/*
+ * OSSL_EVENT_SUBSCRIBER_CLOSURE implements a closure simile, speciflcally
+ * for our event callbacks, to carry along the callback specific data pointer.
+ */
+typedef struct ossl_event_subscriber_closure_st {
+    ossl_event_callback_fn *cb;
+    void *cbarg;
+} OSSL_EVENT_SUBSCRIBER_CLOSURE;
+
+/*
+ * Opaque definition, to allow an optimal internal implementation without
+ * having to change anything else.
+ * This structure is expected to have the semantics of an array of stacks
+ * of events, and should only be affected by event addition, event removal
+ * and event dispatching, all of is internal code in crypto/event/event.c.
+ *
+ * Each element in this implied array represent a priority level, and each
+ * stack (element in this array) simply stores the events for its priority
+ * level in a FIFO manner; events get pushed in and shifted out.
+ */
+struct ossl_events_stacks_st;
+
+struct ossl_event_queue_method_st;
+struct ossl_event_queue_st {
+    const struct ossl_event_queue_method_st *method;
+    void *method_data;
+
+    /*
+     * The number of subscribers determine how many subscribers are listening
+     * to events passing through this queue.  The number of subscribers may
+     * change dynamically.
+     */
+    STACK_OF(OSSL_EVENT_SUBSCRIBER_CLOSURE) *subscribers;
+
+    /*
+     * The array of stacks of events.
+     *
+     * It's possible to have no events stack at all, i.e. for this field to
+     * be NULL.  In that case, the event publishing function will dispatch
+     * events to the subscribers immediately.
+     *
+     * The size of the implied array (i.e. the number of priority levels) is
```

Comment:
Why wouldn't a priority queue be used instead of an array of queues?  This would simplify things and allow for new priorities to be introduced easily, it also allows gaps in priorities (so, e.g. we could have a _highest_ priority that was far apart from the normal priorities (e.g. normal 1-10, highest 10⁶), it also lets us go negative when we realise we missed something).

Using a priority queue would enable us to unify the timer and event queues which would be a big simplification.

The downside to a priority queue is the O(log n) insert/delete time.  I doubt it's a big concern & we can always optimise later.

Reply:
> The downside to a priority queue is the O(log n) insert/delete time

That's exactly the reason that I didn't go with the single queue idea.

Regarding rearranging priorities, I'm not sure what you're getting at.  If we have a set of defined priorities, say like this:

``` C
#define EVENT_PRIO_HIGH 0
#define EVENT_PRIO_LOW 1
```

... and then want something in between, this would be the change:

``` diff
 #define EVENT_PRIO_HIGH 0
-#define EVENT_PRIO_LOW 1
+#define EVENT_PRIO_MEDIUM 1
+#define EVENT_PRIO_LOW 2
```

How hard was that on a scale?

It's not like we're going to change the priority levels live, right?  Or were you imagining dynamic priority levels?

Reply:
That doesn't scale well.  Sure it works while things are tiny but they never stay that way.

There is a good argument for always being about to insert a new priority between any two existing priorities.  That's not possibly with fixed values.  It either requires lots of gaps or reals.

Reply:
So you do see priority levels changing dynamically...  care to elaborate with an example or a scenario?

Reply:
As an extreme: define priorities 1 through 50.  Something new comes up that should be priority between 4 and 5.  Lots of renumbering is required.  Leaving gaps alleviates this to some extent.  However, the more gaps (or just unused) slots, the less efficient as array is.  The break-even with an O(log n) alternative won't be big.  Scanning for the next event requires walking the array of priorities after all.

Reply:
Also remember that the complexity for a lookup into a bucket sorted array used in an event queue isn't O(1).  It's O(k), where k is the number of buckets.  That's O(n k) to process everything.  It doesn't take much for k > log (n), making it less efficient.  Witness the sparse array bucket size reduction I recently did.  We're not sorting a list of numbers and then processing them in order -- the state is dynamic over time.

So bucket sort is good, if we can keep the number of buckets very small.

Reply:
There is no requirement for a single priority queue -- we could have one for events that fire immediately and one for timers.  When processing, check the former before the latter.  Hopefully, the former will always remain short since events are transient short lived things.  Cost wise it would be low.

Reply:
> There is no requirement for a single priority queue -- we could have one for events that fire immediately and one for timers. When processing, check the former before the latter. Hopefully, the former will always remain short since events are transient short lived things. Cost wise it would be low.

Wouldn't that be exactly the same as my array of event lists?  Just that, instead of having an array inside one `ossl_event_queue`, we would have (more or less) an array of `ossl_event_queue`?  I'm not sure what we'd gain with that...

Reply:
I've strong doubts that we would have as many priority levels as you anticipate.  Somewhere else, you mentioned OOB as opposed to "normal events", that makes two.